keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,305 @@
+  * Countermeasure: [CSI migration](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190129-csi-migration.md) and this KEP should graduate together.
+
+* This KEP depends on [CSI migration library](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/csi-translation-lib) ability to handle in-line in-tree volumes. Scheduler will need to get CSI driver name + `VolumeHandle` from them to count them towards the limit.
+
+## Design Details
+
+Existing feature gate `AttachVolumeLimit` will be re-used for implementation of this KEP. The feature is already beta and is enabled by default.
+
+### Test Plan
+
+* [Scheduler benchmark](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-scheduling/scheduler_benchmarking.md) must be extended to run pods with volumes as part of this KEP. Following matrix will be tested:
+  * Predicates:
+    * All volume predicates enabled.
+    * Only deprecated `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount` predicates enabled.
+    * Only `MaxCSIVolumeCountPred` predicate enabled.
+  * API objects:
+    * Both CSINode and Node containing `spec/status.allocatable` for a volume plugin (to simulate kubelet during deprecation period).
+    * Only CSINode containing `spec.drivers[xyz].allocatable` for a volume plugin (to simulate kubelet after deprecation period).
+    * Only Node containing `status.allocatable` for a volume plugin (to simulate old kubelet).
+  * Test results should be ideally the same as before the KEP.
+    * Any deviation needs to be approved by sig-scheduling.
+
+* Run e2e tests and kubelet version skew tests to check that scheduler picks the right values from CSINode or Node.
+
+* Add e2e test that runs pods with both in-tree volumes and CSI driver for the same storage backend and check that they share the same volume limits.
+
+### Graduation Criteria
+
+##### Alpha -> Beta Graduation
+
+N/A (`AttachVolumeLimit` feature is already beta).
+
+##### Beta -> GA Graduation
+
+It must graduate together with CSI migration.
+
+### Upgrade / Downgrade / Version Skew Strategy
+
+During upgrade, downgrade or version skew, kubelet may be older that scheduler. Kubelet will not fill `CSINode.spec` with volume limits and it will fill volume limits into `Node.status`. Scheduler must fall back to `Node.status` when `CSINode` is not available or its `spec` does not contain a volume plugin / CSI driver.
+
+#### Interaction with CSI migration
+
+In-tree volume plugins are being migrated to CSI in a [separate KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190129-csi-migration.md). This section covers possible situations:
+
+* CSI migration feature is `off` for a volume plugin and CSI driver is not installed on a node:
+  * In-tree volume plugin is used for in-tree PVs.
```

Comment:
Not just PVs but inline volumes too

Reply:
ditto for the other sections

Reply:
I have updated this line. Although the proposal does say - we are using "PVs" for brevity when we mean both PVs and inline. 

> For brevity, in-line volumes in pods are handled the same as PVs in all cases above. This assumes that CSI migration can handle in-line volumes!